src/twisted/web/static.py

```diff
@@ -297,12 +299,15 @@ def getChild(self, path, request):
             if fpath is None:
                 return self.childNotFound
 
+        extension = fpath.splitext()[1]
+        if not isinstance(extension, StringType):
+            extension = extension.decode(sys.getfilesystemencoding())
```

Comment:
In a previous review, I questioned the choice of UTF-8 here, because I though `fpath`'s type was determined by the file system's encoding.  I was wrong!

`fpath` is a child path of `site.File`, which in turn is a subclass of `twisted.python.filepath.FilePath`.  `FilePath`s [can represent either `unicode` or `bytes` paths, and accessing a `bytes` child from a `unicode` path results in a `bytes` path](http://twistedmatrix.com/documents/17.9.0/api/twisted.python.filepath.FilePath.html):

```
Python 3.6.1 (v3.6.1:69c0db5050, Apr  3 2017, 01:12:50) 
[GCC 6.3.0 20170221] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from twisted.python.filepath import FilePath
>>> path = FilePath('/unicode/path')
>>> path
FilePath('/unicode/path')
>>> path.child(b"somebytes")
FilePath(b'/unicode/path/somebytes')
```

That's exactly what's happening in this code when you run `twist web --path=/some/path`; `self.path` is `str`, because Python 3 decodes `sys.argv` according to your locale.

However, here in `static.File.getChild`, `fpath` is constructed by calling `FilePath.child` or `FilePath.preauthChild` on the `bytes` object `path`:

https://github.com/twisted/twisted/blob/73961416385261fc876bf8f03cc868424890a689/src/twisted/web/static.py#L289-L293

`path` is `bytes` because it's a _URL path_, and _generally_ [these are encoded with UTF-8](https://tools.ietf.org/html/rfc3986#section-2.5).

I now think the better solution is to do this:

```python
def getChild(self, path, request):
    if isinstance(self.path, twisted.python.compat.unicode):
        try:
            path = path.decode('utf-8')
        except UnicodeDecodeError:
            self._logger.failure("Could not decode {path!r} as UTF-8", path=path)
            return self.childNotFound  # or, better yet, return a 400
    ....
```

Since URL paths are usually UTF-8, this prevents `self.child` or `self.preauthChild` from decoding it with whatever `sys.getfilesystemencoding` returns.  It should also be backwards compatible.